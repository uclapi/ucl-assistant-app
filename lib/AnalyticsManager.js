import * as Amplitude from 'expo-analytics-amplitude'
import Constants from 'expo-constants'

const AMPLITUDE_API_KEY = Constants.manifest.extra
  && Constants.manifest.extra.amplitude
  && Constants.manifest.extra.amplitude.apiKey
  ? Constants.manifest.extra.amplitude.apiKey
  : null

const initialise = () => Amplitude.initialize(AMPLITUDE_API_KEY)

const setUserId = (userId) => {
  // make sure this is unique
  if (userId) {
    Amplitude.setUserId(userId)
  }
}

const setUserProperties = ({ ...userProperties }) => {
  Amplitude.setUserProperties({ ...userProperties })
}

const clearUserProperties = () => {
  Amplitude.clearUserProperties()
}

const logEvent = (eventName, eventProperties) => {
  if (eventProperties) {
    Amplitude.logEventWithProperties(eventName, eventProperties)
  } else {
    Amplitude.logEvent(eventName)
  }
}

const logScreenView = (screenName) => logEvent(`VIEW_SCREEN_${screenName}`)

const events = {
  NOTIFICATIONS_ENABLE: `NOTIFICATIONS_ENABLE`,
  NOTIFICATIONS_SKIP: `NOTIFICATIONS_SKIP`,
  PUSH_NOTIFICATIONS_REGISTER: `PUSH_NOTIFICATIONS_REGISTER`,
  SETTINGS_GIVE_FEEDBACK: `SETTINGS_GIVE_FEEDBACK`,
  SETTINGS_RATE_APP: `SETTINGS_RATE_APP`,
  SETTINGS_VIEW_FAQS: `SETTINGS_VIEW_FAQS`,
}

const AnalyticsManager = {
  clearUserProperties,
  events,
  initialise,
  logEvent,
  logScreenView,
  setUserId,
  setUserProperties,
}

export default (AMPLITUDE_API_KEY && !__DEV__)
  ? AnalyticsManager
  : (
    new Proxy({}, {
      get() {
        return () => { } // do nothing in development mode
      },
    })
  )
