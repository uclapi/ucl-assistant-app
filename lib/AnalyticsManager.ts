import * as Amplitude from 'expo-analytics-amplitude'
import Constants from 'expo-constants'

import configureStore from "../configureStore"

const { store } = configureStore

const AMPLITUDE_API_KEY = Constants.manifest.extra
  && Constants.manifest.extra.amplitude
  && Constants.manifest.extra.amplitude.apiKey
  ? Constants.manifest.extra.amplitude.apiKey
  : null

const initialise = (): Promise<void> => Amplitude.initializeAsync(AMPLITUDE_API_KEY)

const shouldTrackAnalytics = (): boolean => (
  AMPLITUDE_API_KEY
  && !__DEV__
  && store.getState().user.settings.shouldTrackAnalytics
)

const setUserId = (userId): void => {
  if (!shouldTrackAnalytics()) { return }

  // make sure this is unique
  if (userId) {
    Amplitude.setUserIdAsync(userId)
  }
}

const setUserProperties = ({ ...userProperties }): void => {
  if (!shouldTrackAnalytics()) { return }

  Amplitude.setUserPropertiesAsync({ ...userProperties })
}

const clearUserProperties = (): void => {
  Amplitude.clearUserPropertiesAsync()
}

const logEvent = (eventName: string, eventProperties?: unknown): void => {
  if (!shouldTrackAnalytics()) { return }

  if (eventProperties) {
    Amplitude.logEventWithPropertiesAsync(eventName, eventProperties)
  } else {
    Amplitude.logEventAsync(eventName)
  }
}

const logScreenView = (screenName: string): void => {
  if (!shouldTrackAnalytics()) { return }

  logEvent(`VIEW_SCREEN_${screenName}`)
}

const events = {
  NOTIFICATIONS_ENABLE: `NOTIFICATIONS_ENABLE`,
  NOTIFICATIONS_SKIP: `NOTIFICATIONS_SKIP`,
  PUSH_NOTIFICATIONS_REGISTER: `PUSH_NOTIFICATIONS_REGISTER`,
  SETTINGS_GIVE_FEEDBACK: `SETTINGS_GIVE_FEEDBACK`,
  SETTINGS_RATE_APP: `SETTINGS_RATE_APP`,
  SETTINGS_VIEW_FAQS: `SETTINGS_VIEW_FAQS`,
}

export const AnalyticsManager = {
  clearUserProperties,
  events,
  initialise,
  logEvent,
  logScreenView,
  setUserId,
  setUserProperties,
}

export default AnalyticsManager
