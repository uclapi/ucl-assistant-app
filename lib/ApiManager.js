import axios from "axios";
import moment from "moment";
import { encode as btoa } from "base-64";
import {
  ROOMS_URL,
  WORKSPACES_URL,
  EQUIPMENT_URL,
  ROOMBOOKINGS_URL,
} from "../constants/API";

const roomsController = {
  search: async (token = null, query) => {
    if (query && query.length < 3) {
      return {};
    }
    try {
      const results = await axios.get(ROOMS_URL, {
        params: {
          query,
        },
        headers: {
          authorization: `Bearer ${token}`,
          "uclapi-roombookings-version": "1",
        },
      });
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error);
      }
      return results.data.content.rooms;
    } catch (error) {
      throw error;
    }
  },
  getEquipment: async (token = null, { roomid, siteid }) => {
    if (!roomid || !siteid) {
      throw new Error("Must specify roomid and siteid");
    }
    try {
      const results = await axios.get(EQUIPMENT_URL, {
        params: {
          roomid,
          siteid,
        },
        headers: {
          authorization: `Bearer ${token}`,
        },
      });
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error);
      }
      return results.data.content.equipment;
    } catch (error) {
      throw error;
    }
  },
  getBookings: async (
    token = null,
    { roomid, siteid, date = moment().format("YYYYMMDD") },
  ) => {
    if (!roomid || !siteid) {
      throw new Error("Must specify roomid and siteid");
    }
    try {
      const results = await axios.get(ROOMBOOKINGS_URL, {
        params: {
          roomid,
          siteid,
          date,
        },
        headers: {
          authorization: `Bearer ${token}`,
        },
      });
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error);
      }
      return results.data.content.bookings;
    } catch (error) {
      throw error;
    }
  },
};

const workspacesApi = axios.create({
  baseURL: WORKSPACES_URL,
});

const workspacesController = {
  getLiveImage: async (token = null, { surveyId, mapId }) => {
    try {
      const result = await workspacesApi.get("/getliveimage/map.svg", {
        params: {
          survey_id: surveyId,
          map_id: mapId,
        },
        headers: {
          authorization: `Bearer ${token}`,
        },
        responseType: "arraybuffer",
      });
      const image = btoa(
        new Uint8Array(result.data).reduce(
          (data, byte) => data + String.fromCharCode(byte),
          "",
        ),
      );
      // there may be no need to base64 SVGs
      // but I have yet to test this extensively
      // https://css-tricks.com/probably-dont-base64-svg/
      return `data:image/svg+xml;base64,${image}`;
    } catch (error) {
      throw error;
    }
  },
};

export default {
  rooms: roomsController,
  workspaces: workspacesController,
};
