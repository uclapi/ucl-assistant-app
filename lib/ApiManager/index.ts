import Constants from 'expo-constants'
import PeopleController from './PeopleController'
import RoomsController from './RoomsController'
import TimetableController from './TimetableController'
import WorkspacesController from './WorkspacesController'

const clientId = Constants.manifest.extra?.uclapi?.clientId

export default {
  people: PeopleController,
  rooms: RoomsController,
  timetable: TimetableController,
  workspaces: WorkspacesController,
  clientId,
}
