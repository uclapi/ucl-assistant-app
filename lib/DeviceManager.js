import * as Device from 'expo-device'

const isRealDevice = () => Device.isRealDevice

export default {
  isRealDevice,
}
