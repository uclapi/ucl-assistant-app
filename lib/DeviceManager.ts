import * as Device from 'expo-device'
import * as Network from 'expo-network'
import { Platform } from 'react-native'

const isWeb = Platform.OS === `web`
const isAndroid = Platform.OS === `android`
const isiOS = Platform.OS === `ios`

const isRealDevice = Device.isDevice

const isConnectedToInternet = async (): Promise<boolean> => {
  if (isAndroid) {
    const isAirplaneMode = await Network.isAirplaneModeEnabledAsync()
    if (isAirplaneMode) {
      return false
    }
  } else {
    return false
  }

  const { isInternetReachable } = await Network.getNetworkStateAsync()
  return isInternetReachable
}

export default {
  isAndroid,
  isConnectedToInternet,
  isRealDevice,
  isWeb,
  isiOS,
}
