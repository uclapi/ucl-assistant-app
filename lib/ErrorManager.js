import Constants from 'expo-constants'
import * as Sentry from "sentry-expo"

const initialise = () => {
  if (!__DEV__) {
    const { extra: { sentry: { dsn } = {} } = {} } = Constants.manifest
    Sentry.init({
      dsn,
    })

    const { revisionId } = Constants.manifest
    if (typeof revisionId === `string` && revisionId.length > 0) {
      Sentry.setRelease(Constants.manifest.revisionId)
    }
  }
}

const captureError = (error) => {
  if (!__DEV__) {
    Sentry.captureException(error)
  } else {
    console.error(error)
  }
}

export default {
  captureError,
  initialise,
}
