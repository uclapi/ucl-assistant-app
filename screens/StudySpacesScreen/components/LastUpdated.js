import PropTypes from "prop-types"
import React from 'react'
import { momentObj } from "react-moment-proptypes"
import { StyleSheet } from 'react-native'

import { BodyText, ErrorText } from "../../../components/Typography"
import { LocalisationManager } from '../../../lib'

const styles = StyleSheet.create({
  error: {
    marginBottom: 10,
  },
})

class LastUpdated extends React.Component {
  static propTypes = {
    lastModified: PropTypes.oneOfType([momentObj, PropTypes.string]),
  }

  static defaultProps = {
    lastModified: null,
  }

  renderLoading = () => <BodyText>Loading...</BodyText>

  renderError = () => (
    <ErrorText containerStyle={styles.error}>
      Our studyspace data is stale. Sorry about that.
      We&apos;re working on getting this fixed as quickly as possible.
    </ErrorText>
  )

  render() {
    const { lastModified } = this.props
    if (lastModified === null || typeof lastModified !== `object`) {
      return this.renderLoading()
    }

    const isStale = lastModified.isBefore(
      LocalisationManager.getMoment().subtract(5, `minutes`),
    )

    return (
      <>
        {
          isStale ? this.renderError() : null
        }
        <BodyText>
          {`Last updated ${lastModified.calendar().toLowerCase()}`}
        </BodyText>
      </>
    )
  }
}

export default LastUpdated
