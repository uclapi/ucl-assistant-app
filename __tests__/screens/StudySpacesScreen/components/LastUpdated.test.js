/**
 * @jest-environment jsdom
 */
import "react-native"

import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import MockDate from 'mockdate'
import React from 'react'

import { LocalisationManager } from '../../../../lib'
import LastUpdated from
  "../../../../screens/StudySpacesScreen/components/LastUpdated"

Enzyme.configure({ adapter: new Adapter() })

describe(`LastUpdated`, () => {
  MockDate.set(new Date(`2019-11-18T08:47:21`))

  it(`renders the LastUpdated component`, () => {
    const mockProps = {
      lastModified: LocalisationManager.getMoment().subtract(10, `minutes`),
    }
    const wrapper = shallow(<LastUpdated {...mockProps} />)
    expect(wrapper).toMatchSnapshot()
  })

  it(`renders the LastUpdated component loading`, () => {
    const mockProps = {
      lastModified: `${LocalisationManager.parseToDate()}`,
    }
    const wrapper = shallow(<LastUpdated {...mockProps} />)
    expect(wrapper).toMatchSnapshot()
  })

  it(`shows error when studyspace info is stale`, async () => {
    const mockProps = {
      lastModified: LocalisationManager
        .getMoment()
        .subtract(10, `hours`),
    }
    const wrapper = shallow(<LastUpdated {...mockProps} />)
    expect(wrapper).toMatchSnapshot()
  })
})
