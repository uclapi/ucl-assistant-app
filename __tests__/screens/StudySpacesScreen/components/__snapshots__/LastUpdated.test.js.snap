// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LastUpdated renders the LastUpdated component 1`] = `
<Fragment>
  <BodyText
    style={Object {}}
  >
    Last updated today at 8:43 am
  </BodyText>
</Fragment>
`;

exports[`LastUpdated renders the LastUpdated component loading 1`] = `
<BodyText
  style={Object {}}
>
  Loading...
</BodyText>
`;

exports[`LastUpdated shows error when studyspace info is stale 1`] = `
<Fragment>
  <ErrorText
    containerStyle={
      Object {
        "marginBottom": 10,
      }
    }
    style={Object {}}
  >
    Our studyspace data is stale. Sorry about that. We're working on getting this fixed as quickly as possible.
  </ErrorText>
  <BodyText
    style={Object {}}
  >
    Last updated today at 8:41 am
  </BodyText>
</Fragment>
`;
